// @primer/react/lib-esm/drafts/MarkdownEditor/_ToolbarButton.js
import React, { forwardRef, useContext } from 'react';
import { IconButton } from '../../Button/IconButton.js';
import '../../Box/Box.js';
import '@styled-system/css';
import 'deepmerge';
import '../../ThemeProvider.js';
import '../../Button/types.js';
import '../../utils/defaultSxProp.js';
import '../../CounterLabel/CounterLabel.js';
import '../../Button/Button.js';
import { MarkdownEditorContext } from './_MarkdownEditorContext.js';

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const ToolbarButton = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    disabled,
    condensed
  } = useContext(MarkdownEditorContext);
  return /*#__PURE__*/React.createElement(IconButton, _extends({
    ref: ref,
    size: condensed ? 'small' : 'medium',
    variant: "invisible",
    disabled: disabled
    // Prevent focus leaving input:
    ,
    onMouseDown: e => e.preventDefault()
  }, props, {
    sx: {
      color: 'fg.muted',
      ...props.sx
    }
  }));
});
ToolbarButton.displayName = 'MarkdownEditor.ToolbarButton';

export { ToolbarButton };
