// @primer/react/lib/CounterLabel/CounterLabel.js
'use strict';

var React = require('react');
var Box = require('../Box/Box.js');
require('@styled-system/css');
var merge = require('deepmerge');
var _VisuallyHidden = require('../_VisuallyHidden.js');
var defaultSxProp = require('../utils/defaultSxProp.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);
var merge__default = /*#__PURE__*/_interopDefault(merge);

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const CounterLabel = /*#__PURE__*/React.forwardRef(({
  scheme = 'secondary',
  sx = defaultSxProp.defaultSxProp,
  children,
  ...props
}, forwardedRef) => {
  return /*#__PURE__*/React__default.default.createElement(React__default.default.Fragment, null, /*#__PURE__*/React__default.default.createElement(Box, _extends({
    "aria-hidden": "true",
    sx: merge__default.default({
      display: 'inline-block',
      padding: '2px 5px',
      fontSize: 0,
      fontWeight: 'bold',
      lineHeight: 'condensedUltra',
      borderRadius: '20px',
      backgroundColor: scheme === 'primary' ? 'neutral.emphasis' : 'neutral.muted',
      color: scheme === 'primary' ? 'fg.onEmphasis' : 'fg.default',
      '&:empty': {
        display: 'none'
      }
    }, sx)
  }, props, {
    as: "span",
    ref: forwardedRef
  }), children), /*#__PURE__*/React__default.default.createElement(_VisuallyHidden, null, "\xA0(", children, ")"));
});
CounterLabel.displayName = 'CounterLabel';

module.exports = CounterLabel;
