// @primer/react/lib-esm/Button/ButtonBase.js
import React, { forwardRef, useMemo } from 'react';
import Box from '../Box/Box.js';
import '@styled-system/css';
import merge from 'deepmerge';
import { useTheme } from '../ThemeProvider.js';
import { StyledButton } from './types.js';
import { getButtonStyles, getVariantStyles, getAlignContentSize } from './styles.js';
import { useRefObjectAsForwardedRef } from '../hooks/useRefObjectAsForwardedRef.js';
import { defaultSxProp } from '../utils/defaultSxProp.js';
import CounterLabel from '../CounterLabel/CounterLabel.js';

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const ButtonBase = /*#__PURE__*/forwardRef(({
  children,
  as: Component = 'button',
  sx: sxProp = defaultSxProp,
  ...props
}, forwardedRef) => {
  const {
    leadingVisual: LeadingVisual,
    trailingVisual: TrailingVisual,
    trailingAction: TrailingAction,
    count,
    icon: Icon,
    variant = 'default',
    size = 'medium',
    alignContent = 'center',
    block = false,
    inactive,
    ...rest
  } = props;
  const innerRef = React.useRef(null);
  useRefObjectAsForwardedRef(forwardedRef, innerRef);
  const {
    theme
  } = useTheme();
  const baseStyles = useMemo(() => {
    return merge.all([getButtonStyles(theme), getVariantStyles(variant, theme)]);
  }, [theme, variant]);
  const sxStyles = useMemo(() => {
    return merge(baseStyles, sxProp);
  }, [baseStyles, sxProp]);
  const iconWrapStyles = {
    display: 'flex',
    pointerEvents: 'none'
  };
  if (process.env.NODE_ENV !== "production") {
    /**
     * The Linter yells because it thinks this conditionally calls an effect,
     * but since this is a compile-time flag and not a runtime conditional
     * this is safe, and ensures the entire effect is kept out of prod builds
     * shaving precious bytes from the output, and avoiding mounting a noop effect
     */
    // eslint-disable-next-line react-hooks/rules-of-hooks
    React.useEffect(() => {
      if (innerRef.current && !(innerRef.current instanceof HTMLButtonElement) && !(innerRef.current instanceof HTMLAnchorElement)) {
        // eslint-disable-next-line no-console
        console.warn('This component should be an instanceof a semantic button or anchor');
      }
    }, [innerRef]);
  }
  return /*#__PURE__*/React.createElement(StyledButton, _extends({
    as: Component,
    sx: sxStyles
  }, rest, {
    ref: innerRef,
    "data-block": block ? 'block' : null,
    "data-size": size === 'small' || size === 'large' ? size : undefined,
    "data-no-visuals": !LeadingVisual && !TrailingVisual && !TrailingAction ? true : undefined,
    "data-inactive": inactive ? true : undefined
  }), Icon ? /*#__PURE__*/React.createElement(Icon, null) : /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Box, {
    as: "span",
    "data-component": "buttonContent",
    sx: getAlignContentSize(alignContent)
  }, LeadingVisual && /*#__PURE__*/React.createElement(Box, {
    as: "span",
    "data-component": "leadingVisual",
    sx: {
      ...iconWrapStyles
    }
  }, /*#__PURE__*/React.createElement(LeadingVisual, null)), children && /*#__PURE__*/React.createElement("span", {
    "data-component": "text"
  }, children), count !== undefined && !TrailingVisual ? /*#__PURE__*/React.createElement(Box, {
    as: "span",
    "data-component": "trailingVisual",
    sx: {
      ...iconWrapStyles
    }
  }, /*#__PURE__*/React.createElement(CounterLabel, {
    "data-component": "ButtonCounter"
  }, count)) : TrailingVisual ? /*#__PURE__*/React.createElement(Box, {
    as: "span",
    "data-component": "trailingVisual",
    sx: {
      ...iconWrapStyles
    }
  }, /*#__PURE__*/React.createElement(TrailingVisual, null)) : null), TrailingAction && /*#__PURE__*/React.createElement(Box, {
    as: "span",
    "data-component": "trailingAction",
    sx: {
      ...iconWrapStyles
    }
  }, /*#__PURE__*/React.createElement(TrailingAction, null))));
});

export { ButtonBase };
