// @primer/react/lib/internal/components/ScrollableRegion.js
'use strict';

var React = require('react');
var Box = require('../../Box/Box.js');
var useOverflow = require('../hooks/useOverflow.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function ScrollableRegion({
  'aria-labelledby': labelledby,
  children,
  ...rest
}) {
  const ref = React__default.default.useRef(null);
  const hasOverflow = useOverflow.useOverflow(ref);
  const regionProps = hasOverflow ? {
    'aria-labelledby': labelledby,
    role: 'region',
    tabIndex: 0
  } : {};
  return /*#__PURE__*/React__default.default.createElement(Box, _extends({}, rest, regionProps, {
    ref: ref,
    sx: {
      overflow: 'auto'
    }
  }), children);
}
ScrollableRegion.displayName = "ScrollableRegion";

exports.ScrollableRegion = ScrollableRegion;
