// @primer/react/lib/legacy-theme/ts/colors/index.js
'use strict';

var light = require('./light.js');
var light_high_contrast = require('./light_high_contrast.js');
var light_colorblind = require('./light_colorblind.js');
var light_tritanopia = require('./light_tritanopia.js');
var dark = require('./dark.js');
var dark_dimmed = require('./dark_dimmed.js');
var dark_high_contrast = require('./dark_high_contrast.js');
var dark_colorblind = require('./dark_colorblind.js');
var dark_tritanopia = require('./dark_tritanopia.js');

const colors = {
  light,
  // eslint-disable-next-line camelcase
  light_high_contrast: light_high_contrast,
  // eslint-disable-next-line camelcase
  light_colorblind: light_colorblind,
  // eslint-disable-next-line camelcase
  light_tritanopia: light_tritanopia,
  dark,
  dark_dimmed: dark_dimmed,
  // eslint-disable-next-line camelcase
  dark_high_contrast: dark_high_contrast,
  // eslint-disable-next-line camelcase
  dark_colorblind: dark_colorblind,
  // eslint-disable-next-line camelcase
  dark_tritanopia: dark_tritanopia
};

exports.colors = colors;
